### PR TITLE
Add distSQL syntax: 'drop sharding algorithm'.

### PR DESCRIPTION
**Pull request by RaigorJiang**
Add distSQL syntax: 'drop sharding algorithm'.

example: 
```
DROP SHARDING ALGORITHM t_user_mod;

DROP SHARDING ALGORITHM t_order_item_inline,t_order_inline, database_inline;
```

### CODE REVIEW

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/java/org/apache/shardingsphere/sharding/distsql/handler/update/DropShardingAlgorithmStatementUpdater.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    public void checkSQLStatement(final String schemaName, final DropShardingAlgorithmStatement sqlStatement, 
+                                  final ShardingRuleConfiguration currentRuleConfig, final ShardingSphereResource resource) throws RuleDefinitionViolationException {
+        checkCurrentRuleConfiguration(schemaName, currentRuleConfig);
+        checkShardingAlgorithmsExisted(schemaName, sqlStatement, currentRuleConfig);
+        checkShardingAlgorithmsInUsed(schemaName, sqlStatement, currentRuleConfig);
+    }
+    
+    private void checkCurrentRuleConfiguration(final String schemaName, final ShardingRuleConfiguration currentRuleConfig) throws RequiredAlgorithmMissedException {
+        if (null == currentRuleConfig) {
+            throw new RequiredAlgorithmMissedException(schemaName);
```
Comment (menghaoranss) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/java/org/apache/shardingsphere/sharding/distsql/handler/update/DropShardingAlgorithmStatementUpdater.java:47]: RequiredRuleMissedException should be thrown here

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/java/org/apache/shardingsphere/sharding/distsql/handler/update/DropShardingAlgorithmStatementUpdater.java [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        Collection<String> allInUsed = getAllOfAlgorithmsInUsed(currentRuleConfig);
+        Collection<String> usedAlgorithms = sqlStatement.getAlgorithmNames().stream().filter(allInUsed::contains).collect(Collectors.toList());
+        if (!usedAlgorithms.isEmpty()) {
+            throw new AlgorithmInUsedException(schemaName, usedAlgorithms);
+        }
+    }
+    
+    private Collection<String> getAllOfAlgorithmsInUsed(final ShardingRuleConfiguration shardingRuleConfig) {
+        Collection<String> result = new LinkedHashSet<>();
+        shardingRuleConfig.getTables().stream().forEach(each -> {
```
Comment (menghaoranss) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/java/org/apache/shardingsphere/sharding/distsql/handler/update/DropShardingAlgorithmStatementUpdater.java:71]: autoTables in ShardingRuleConfiguration should also be added

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/java/org/apache/shardingsphere/sharding/distsql/handler/update/DropShardingAlgorithmStatementUpdater.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+        checkShardingAlgorithmsInUsed(schemaName, sqlStatement, currentRuleConfig);
+    }
+    
+    private void checkCurrentRuleConfiguration(final String schemaName, final ShardingRuleConfiguration currentRuleConfig) throws RequiredAlgorithmMissedException {
+        if (null == currentRuleConfig) {
+            throw new RequiredAlgorithmMissedException(schemaName);
+        }
+    }
+    
+    private void checkShardingAlgorithmsExisted(final String schemaName, final DropShardingAlgorithmStatement sqlStatement,
```
Comment (menghaoranss) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/java/org/apache/shardingsphere/sharding/distsql/handler/update/DropShardingAlgorithmStatementUpdater.java:51]: checkShardingAlgorithmsIsExisted is better?

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/java/org/apache/shardingsphere/sharding/distsql/handler/update/DropShardingAlgorithmStatementUpdater.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+                                                final ShardingRuleConfiguration currentRuleConfig) throws RequiredAlgorithmMissedException {
+        Collection<String> currentShardingAlgorithms = getCurrentShardingAlgorithms(currentRuleConfig);
+        Collection<String> notExistedAlgorithms = sqlStatement.getAlgorithmNames().stream().filter(each -> !currentShardingAlgorithms.contains(each)).collect(Collectors.toList());
+        if (!notExistedAlgorithms.isEmpty()) {
+            throw new RequiredAlgorithmMissedException(schemaName, notExistedAlgorithms);
+        }
+    }
+    
+    private void checkShardingAlgorithmsInUsed(final String schemaName, final DropShardingAlgorithmStatement sqlStatement, final ShardingRuleConfiguration currentRuleConfig)
```
Comment (menghaoranss) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/java/org/apache/shardingsphere/sharding/distsql/handler/update/DropShardingAlgorithmStatementUpdater.java:60]: checkShardingAlgorithmsIsInUsed

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/java/org/apache/shardingsphere/sharding/distsql/handler/update/DropShardingAlgorithmStatementUpdater.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    private Collection<String> getCurrentShardingAlgorithms(final ShardingRuleConfiguration shardingRuleConfig) {
+        return shardingRuleConfig.getShardingAlgorithms().keySet();
+    }
+    
+    @Override
+    public boolean updateCurrentRuleConfiguration(final DropShardingAlgorithmStatement sqlStatement, final ShardingRuleConfiguration currentRuleConfig) {
+        for (String each : sqlStatement.getAlgorithmNames()) {
+            dropShardingAlgorithm(currentRuleConfig, each);
+        }
+        return false;
```
Comment (menghaoranss) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/java/org/apache/shardingsphere/sharding/distsql/handler/update/DropShardingAlgorithmStatementUpdater.java:91]: should real result be returned here?

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-parser/src/main/antlr4/imports/sharding/RDLStatement.g4 [hunk 6]
```diff
@@ -54,6 +54,10 @@ dropShardingBindingTableRules
 dropShardingBroadcastTableRules
     : DROP SHARDING BROADCAST TABLE RULES
     ;
+    
+dropShardingAlgorithm
+    : DROP SHARDING ALGORITHM algorithmName (COMMA algorithmName)*
```
Comment (menghaoranss) [inline shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-parser/src/main/antlr4/imports/sharding/RDLStatement.g4:59]: Does we need support `from schema`?

**Review by menghaoranss — CHANGES_REQUESTED** [review]

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/java/org/apache/shardingsphere/sharding/distsql/handler/update/DropShardingAlgorithmStatementUpdater.java (hunk 1 above)
Comment (RaigorJiang) [reply shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/java/org/apache/shardingsphere/sharding/distsql/handler/update/DropShardingAlgorithmStatementUpdater.java:47]: 'shardingAlgorithms'  is part of the ‘ShardingRuleConfiguration’, it is not a rule itself, so I created RequiredAlgorithmMissedException.

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/java/org/apache/shardingsphere/sharding/distsql/handler/update/DropShardingAlgorithmStatementUpdater.java (hunk 2 above)
Comment (RaigorJiang) [reply shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/java/org/apache/shardingsphere/sharding/distsql/handler/update/DropShardingAlgorithmStatementUpdater.java:71]: Good idea 👍

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/java/org/apache/shardingsphere/sharding/distsql/handler/update/DropShardingAlgorithmStatementUpdater.java (hunk 3 above)
Comment (RaigorJiang) [reply shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/java/org/apache/shardingsphere/sharding/distsql/handler/update/DropShardingAlgorithmStatementUpdater.java:51]: This is a plural form, and ‘is’ is not suitable. If necessary, I prefer ‘checkShardingAlgorithmsIsExist’,  similar to ’File.exists‘

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/java/org/apache/shardingsphere/sharding/distsql/handler/update/DropShardingAlgorithmStatementUpdater.java (hunk 4 above)
Comment (RaigorJiang) [reply shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/java/org/apache/shardingsphere/sharding/distsql/handler/update/DropShardingAlgorithmStatementUpdater.java:60]: Plural form too.

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/java/org/apache/shardingsphere/sharding/distsql/handler/update/DropShardingAlgorithmStatementUpdater.java (hunk 5 above)
Comment (RaigorJiang) [reply shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/java/org/apache/shardingsphere/sharding/distsql/handler/update/DropShardingAlgorithmStatementUpdater.java:91]: refer to `DropShardingTableRuleStatementUpdater` and `DropShardingBindingTableRuleStatementUpdater`

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-parser/src/main/antlr4/imports/sharding/RDLStatement.g4 (hunk 6 above)
Comment (RaigorJiang) [reply shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-parser/src/main/antlr4/imports/sharding/RDLStatement.g4:59]: I don't think so.   'Sharding Algorithm' is part of sharding rule, and 'Drop Sharing Rule' does not support from schema yet.

File: shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/java/org/apache/shardingsphere/sharding/distsql/handler/update/DropShardingAlgorithmStatementUpdater.java (hunk 3 above)
Comment (RaigorJiang) [reply shardingsphere-features/shardingsphere-sharding/shardingsphere-sharding-distsql/shardingsphere-sharding-distsql-handler/src/main/java/org/apache/shardingsphere/sharding/distsql/handler/update/DropShardingAlgorithmStatementUpdater.java]: Finally, I change it to `checkToBeDroppedShardingAlgorithms`, like the `checkToBeDroppedShardingTableNames` int `DropShardingTableRuleStatementUpdater`, is it OK?

**Review by menghaoranss — APPROVED** [review]

**Pull request merged**